### ISSUE TITLE
If we rely on crossbeam_channel, we can have a 'sync' Writer thats similar to the MpscWriter

### ISSUE THREAD

**Issue by museun** [museun/twitchchat#150]
```rust

struct Encoder {
    tx: Tx<Vec<u8>>,
    buf: Vec<u8>,
}

impl Encoder {
    fn new(tx: Tx<Vec<u8>>) -> Self {
        Self {
            tx,
            buf: Vec::new(),
        }
    }
}

impl Write for Encoder {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.buf.extend_from_slice(buf);
        Ok(buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        match self.tx.try_send(std::mem::take(&mut self.buf)) {
            Ok(..) => Ok(()),
            Err(mpsc::TrySendError::Full(buf)) => {
                std::mem::replace(&mut self.buf, buf);
                Err(std::io::Error::new(
                    std::io::ErrorKind::WouldBlock,
                    "cannot write, other half isn't ready",
                ))
            }
            Err(mpsc::TrySendError::Disconnected(..)) => Err(std::io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                "cannot write, other half dropped",
            )),
        }
    }
}
```
a simple implementation of it, basically does the same as the tokio mpsc version, but synchronous.

We could make it feature-gated so people can exclude crossbeam_channel if they don't want, or will use this style of Encoder/Writer

**Comment by museun** [comment]
This should be possible with std channels. But there doesn't seem to be a way to convert a std channel into a crossbeam channel.
A crossbeam channel was way more ergonomic because it supports its `select!{} / Select` types for non-blocking recvs.

A feature could provided to swap out std with crossbeam. We can default the std one.

**Comment by museun** [comment]
closed in 3bb9cac2d41db403129b1d30c7d28c81cfcb0442
